src/codegen/general.rs

```diff
@@ -113,6 +113,91 @@ fn format_parent_name(env: &Env, p: &StatusedTypeId) -> String {
     }
 }
 
+pub fn define_fundamental_type(
+    w: &mut dyn Write,
+    env: &Env,
+    type_name: &str,
+    glib_name: &str,
+    glib_func_name: &str,
+    ref_func: Option<&str>,
+    unref_func: Option<&str>,
+    parents: &[StatusedTypeId],
+) -> Result<()> {
+    let sys_crate_name = env.main_sys_crate_name();
+    writeln!(w, "{} {{", use_glib_type(env, "wrapper!"))?;
+    doc_alias(w, glib_name, "", 1)?;
+    writeln!(
+        w,
+        "\tpub struct {}(Shared<{}::{}>);",
+        type_name, sys_crate_name, glib_name
+    )?;
+    writeln!(w)?;
+
+    writeln!(w, "\tmatch fn {{")?;
+    let (ref_fn, unref_fn, parent_ffi_type) = if parents.is_empty() {
+        // it's the super-type, it must have a ref/unref functions
+        (
+            ref_func.unwrap().to_owned(),
+            unref_func.unwrap().to_owned(),
+            "ptr".to_owned(),
+        )
+    } else {
+        let (ref_fn, unref_fn, parent_c_type): (String, String, String) = parents
+            .iter()
+            .filter_map(|p| {
+                use crate::library::*;
+                let type_ = env.library.type_(p.type_id);
+                match type_ {
+                    Type::Class(class) => Some((
+                        class.ref_fn.as_ref().unwrap().clone(),
+                        class.unref_fn.as_ref().unwrap().clone(),
+                        class.c_type.clone(),
+                    )),
+                    _ => None,
+                }
+            })
+            .next()
+            .unwrap();
+        // otherwise get that information from the parent
+        (
+            ref_fn,
+            unref_fn,
+            format!("ptr as *mut {}::{}", sys_crate_name, parent_c_type),
+        )
```

Comment:
Just noticed thats the sys_crate_name here is wrong, it should get the one of the parent type in case it comes from a different crate